File: 173.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    stack<TreeNode *> s;
    TreeNode *pr;
    BSTIterator(TreeNode *root) {
        while (root) {
            //cout << root->val << endl;
            s.push(root);
            root = root->left;
        }
        pr = root;
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !s.empty() || pr;
    }

    /** @return the next smallest number */
    int next() {
        int v;
        while (pr) {
            s.push(pr);
            pr = pr->left;
        }
        pr = s.top();
        s.pop();
        v = pr->val;
        pr = pr->right;
        return v;
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: 173.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


def bst(root, l):
    if root:
        bst(root.left, l)
        l.append(root.val)
        bst(root.right, l)
    
class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.l = []
        self.idx = 0
        bst(root, self.l)

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.idx < len(self.l)
        

    def next(self):
        """
        :rtype: int
        """
        val = self.l[self.idx]
        self.idx += 1
        return val

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())
